libraries/message/src/descriptor.rs

```diff
@@ -0,0 +1,256 @@
+use crate::{
+    config::{CommunicationConfig, Input, InputMapping, NodeRunConfig},
+    id::{DataId, NodeId, OperatorId},
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_with_expand_env::with_expand_envs;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fmt,
+    path::PathBuf,
+};
+
+pub const SHELL_SOURCE: &str = "shell";
+pub const DYNAMIC_SOURCE: &str = "dynamic";
+
+/// Dataflow description
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
```

Comment:
Same as above about denying unknown fields that we might want to add in the future.

Reply:
Removed as well.